jlog.c

```diff
@@ -1375,17 +1539,69 @@ int jlog_ctx_write_message(jlog_ctx *ctx, jlog_message *mess, struct timeval *wh
     hdr.tv_sec = now.tv_sec;
     hdr.tv_usec = now.tv_usec;
   }
+
+  /* we store the original message size in the header */
   hdr.mlen = mess->mess_len;
-  if (!jlog_file_pwrite(ctx->data, &hdr, sizeof(hdr), current_offset)) {
-    FASSERT(0, "jlog_file_pwrite failed in jlog_ctx_write_message");
-    SYS_FAIL(JLOG_ERR_FILE_WRITE);
+
+  struct iovec v[2];
+  v[0].iov_base = (void *) &hdr;
+  v[0].iov_len = hdr_size;
+
+  if (IS_COMPRESS_MAGIC(ctx)) {
+    if (jlog_compress(mess->mess, mess->mess_len, &v[1].iov_base, (size_t *)&hdr.compressed_len) != 0) {
+      FASSERT(0, "jlog_compress failed in jlog_ctx_write_message");
+      SYS_FAIL(JLOG_ERR_FILE_WRITE);
+    }
+    v[1].iov_len = hdr.compressed_len;
+  } else {
+    v[1].iov_base = mess->mess;
+    v[1].iov_len = mess->mess_len;
+  }
+
+  size_t total_size = v[0].iov_len + v[1].iov_len;
+  if (ctx->pre_commit_pos + total_size > ctx->pre_commit_end) {
+    if ((current_offset = jlog_file_size(ctx->data)) == -1)
+      SYS_FAIL(JLOG_ERR_FILE_SEEK);
+    if(ctx->meta->unit_limit <= current_offset) {
+      jlog_file_unlock(ctx->data);
+      __jlog_close_writer(ctx);
+      __jlog_metastore_atomic_increment(ctx);
+      if (IS_COMPRESS_MAGIC(ctx)) {
+        free(v[1].iov_base);
+      }
+      goto begin;
+    }
+
+    /* we have to flush our pre_commit_buffer out to the real log */
+    if (!jlog_file_pwrite(ctx->data, ctx->pre_commit_buffer, 
+                          ctx->pre_commit_pos - ctx->pre_commit_buffer, 
+                          current_offset)) {
+      FASSERT(0, "jlog_file_pwrite failed in jlog_ctx_write_message");
+      SYS_FAIL(JLOG_ERR_FILE_WRITE);
+    }
+
+    /* rewind the pre_commit_buffer to beginning */
+    ctx->pre_commit_pos = ctx->pre_commit_buffer;
+    /* ensure we save this in the mmapped data */
+    *ctx->pre_commit_pointer = 0;
+  }
+  
+  /**
+   * Write the iovecs to the pre-commit buffer 
+   * 
+   * This is protected by the file lock on the main data file so needs no special treatment
+   */
+  for (i = 0; i < 2; i++) {
```

Comment:
If the message  doesn't fit in the precommit buffer, it should be written immediately (old style)